crates/polars-core/src/chunked_array/logical/datetime.rs

```diff
@@ -31,63 +31,66 @@ impl LogicalType for DatetimeChunked {
+            (Datetime(from_unit, _), Datetime(to_unit, tz)) => {
+                let (multiplier, divisor) = match (from_unit, to_unit) {
+                    // scaling from lower precision to higher precision
+                    (TimeUnit::Milliseconds, TimeUnit::Nanoseconds) => (Some(1_000_000i64), None),
+                    (TimeUnit::Milliseconds, TimeUnit::Microseconds) => (Some(1_000i64), None),
+                    (TimeUnit::Microseconds, TimeUnit::Nanoseconds) => (Some(1_000i64), None),
+                    // scaling from higher precision to lower precision
+                    (TimeUnit::Nanoseconds, TimeUnit::Milliseconds) => (None, Some(1_000_000i64)),
+                    (TimeUnit::Nanoseconds, TimeUnit::Microseconds) => (None, Some(1_000i64)),
+                    (TimeUnit::Microseconds, TimeUnit::Milliseconds) => (None, Some(1_000i64)),
+                    _ => return self.0.cast(dtype),
+                };
+                let result = match multiplier {
+                    // scale to higher precision (eg: ms → us, ms → ns, us → ns)
+                    Some(m) => Ok((self.0.as_ref() * m)
+                        .into_datetime(*to_unit, tz.clone())
+                        .into_series()),
+                    // scale to lower precision (eg: ns → us, ns → ms, us → ms)
+                    None => match divisor {
+                        Some(d) => Ok(self
+                            .0
+                            .apply_values(|v| v.div_euclid(d))
+                            .into_datetime(*to_unit, tz.clone())
+                            .into_series()),
+                        None => unreachable!("must always have a time unit divisor here"),
+                    },
+                };
+                result
             },
             #[cfg(feature = "dtype-date")]
-            (Datetime(tu, _), Date) => match tu {
-                TimeUnit::Nanoseconds => Ok((self.0.as_ref() / NS_IN_DAY)
-                    .cast(&Int32)
-                    .unwrap()
-                    .into_date()
-                    .into_series()),
-                TimeUnit::Microseconds => Ok((self.0.as_ref() / US_IN_DAY)
-                    .cast(&Int32)
-                    .unwrap()
-                    .into_date()
-                    .into_series()),
-                TimeUnit::Milliseconds => Ok((self.0.as_ref() / MS_IN_DAY)
-                    .cast(&Int32)
-                    .unwrap()
-                    .into_date()
-                    .into_series()),
+            (Datetime(tu, _), Date) => {
+                let cast_to_date = |tu_in_day: i64| {
+                    let mut dt = self
+                        .0
+                        .apply_values(|v| v.div_euclid(tu_in_day))
+                        .cast(&Int32)
+                        .unwrap()
+                        .into_date()
+                        .into_series();
+                    dt.set_sorted_flag(self.is_sorted_flag());
```

Comment:
nice. wouldn't downcasting/upcasting the time_unit also preserve sortedness?

Reply:
Yup; we can likely generalise sort-preservation at a higher level (eg: all casts between units/numeric/temporal representations), which would be ideal 👌

Reply:
agree 👍